### PR TITLE
feat(cast): Add integer radix (base) parsing, add --to-base

### PR DESCRIPTION
**Pull request by DaniPopes**
## Motivation

Initially wanted to add a `to_base`, then saw there already was some kind of base parsing implemented for shl/shr, so I seized the opportunity to do some more stuff

## Solution

- added `cast --to-base <number> <base_out> [--base-in <base>]`, which tries to match `<number>` to a base (binary, hex...) by looking at prefix or parsing it directly, and convert it to `<to_base>`, with an optional `--base in` in ambiguous cases, similar to how it is currently done in `SimpleCast::{shl, shr}`
- added `base.rs`, includes:
  - `Base` enum for parsing radix from strings or numbers' bases;
  - `NumberWithBase` struct for parsing numbers and formatting them into bases;
  - `ToBase` trait for easy one function conversion by using the above items;
- ordering/grouping functions: matching a from/to pair where possible, eg. `{from,to}_fix` etc;
- standardizing names: `--to-asci`'s matching `SimpleCast` function was `ascii()` while most others have the same name as the flag;
- standardizing function args: `from_wei`, `to_wei`, `to_unit` all had different args from every other function;
- refactoring some weird hex formatting: in `to_{uint256,int256}`

Any suggestion is welcome!

### CODE REVIEW

**Comment by DaniPopes** [comment]
Failing lint unrelated

File: cast/src/lib.rs [hunk 1]
```diff
@@ -794,55 +742,48 @@ impl SimpleCast {
...
+    ///     assert_eq!(Cast::from_fix(1, "1.0")?, 10.into());
+    ///     assert_eq!(Cast::from_fix(2, "0.10")?, 10.into());
+    ///     assert_eq!(Cast::from_fix(3, "0.010")?, 10.into());
     ///
     ///     Ok(())
     /// }
     /// ```
-    pub fn max_uint() -> Result<U256> {
-        Ok(U256::MAX)
+    pub fn from_fix(decimals: u32, value: &str) -> Result<U256> {
```
Comment (mattsse) [inline cast/src/lib.rs:779]:
```suggestion
    pub fn from_fixed_point(decimals: u32, value: &str) -> Result<U256> {
```

File: cast/src/lib.rs [hunk 2]
```diff
@@ -984,8 +836,32 @@ impl SimpleCast {
...
+    /// # fn main() -> eyre::Result<()> {
+    /// let addr = Address::from_str("0xb7e390864a90b7b923c9f9310c6f98aafe43f707")?;
+    /// let addr = Cast::to_checksum_address(&addr)?;
+    /// assert_eq!(addr, "0xB7e390864a90b7b923C9f9310C6F98aafE43F707");
+    ///
+    /// # Ok(())
+    /// # }
+    /// ```
+    pub fn to_checksum_address(address: &Address) -> Result<String> {
+        Ok(utils::to_checksum(address, None))
```
Comment (mattsse) [inline cast/src/lib.rs:858]: looks like we don't need a result here?

File: cast/src/lib.rs [hunk 3]
```diff
@@ -1030,133 +907,95 @@ impl SimpleCast {
...
+        Ok(match unit {
+            "eth" | "ether" => ethers_core::utils::format_units(value, 18)?
                 .trim_end_matches(".000000000000000000")
                 .to_string(),
             "gwei" | "nano" | "nanoether" => ethers_core::utils::format_units(value, 9)?
                 .trim_end_matches(".000000000")
                 .to_string(),
             "wei" => ethers_core::utils::format_units(value, 0)?.trim_end_matches(".0").to_string(),
-            _ => return Err(eyre::eyre!("invalid unit")),
+            _ => eyre::bail!("invalid unit"),
```
Comment (mattsse) [inline cast/src/lib.rs:941]: let's include the invalid values in the error

File: cast/src/lib.rs [hunk 4]
```diff
@@ -1030,133 +907,95 @@ impl SimpleCast {
...
-    pub fn to_rlp(value: &str) -> Result<String> {
-        let val = serde_json::from_str(value).unwrap_or(serde_json::Value::String(value.parse()?));
-        let item = Item::value_to_item(&val)?;
-        Ok(format!("0x{}", hex::encode(rlp::encode(&item))))
+    pub fn to_wei(value: &str, unit: &str) -> Result<String> {
+        Ok(match unit {
+            "gwei" => ethers_core::utils::parse_units(value, 9),
+            _ => ethers_core::utils::parse_units(value, 18),
+        }?
+        .to_string())
```
Comment (mattsse) [inline cast/src/lib.rs:992]: for readability use two steps here

let wei = match ...
Ok(wei.to_string())

File: cast/src/lib.rs [hunk 5]
```diff
@@ -1224,35 +1101,229 @@ impl SimpleCast {
...
+    /// # }
+    /// ```
+    pub async fn generate_interface(
+        address_or_path: InterfacePath,
+    ) -> Result<Vec<InterfaceSource>> {
+        let (contract_abis, contract_names): (Vec<Abi>, Vec<String>) = match address_or_path {
+            InterfacePath::Local { path, name } => {
+                let file = std::fs::read_to_string(&path).wrap_err("unable to read abi file")?;
+
+                let mut json: serde_json::Value = serde_json::from_str(&file)?;
```
Comment (mattsse) [inline cast/src/lib.rs:1249]: this needs to be rebased after https://github.com/foundry-rs/foundry/pull/3124

File: cli/src/opts/cast.rs [hunk 6]
```diff
@@ -213,8 +203,8 @@ Examples:
         #[clap(allow_hyphen_values = true, value_name = "VALUE")]
         // negative values not yet supported internally
         value: Option<String>,
-        #[clap(value_name = "UNIT")]
-        unit: Option<String>,
+        #[clap(value_name = "UNIT", default_value = "eth")]
+        unit: String,
```
Comment (mattsse) [inline cli/src/opts/cast.rs:207]: good change!

**Review by mattsse — CHANGES_REQUESTED** [review]
this is great!
tysm!

only have a few nits, overall lgtm

**Comment by mattsse** [comment]
thanks!

the failing test is bcause checksum address no longer returns an result, just need to remove the unwrap

**Comment by mattsse** [comment]
> I believe failing tests are unrelated
> 
> what do we think of deprecating `--to-dec N` and `--to-hex N` in favor of `--to-base N dec/hex`? or should we just keep them as aliases and hint to `--to-base` in the help like in `--{to,from}wei`?

let's just keep them as clap `visible_alias`

**Comment by DaniPopes** [comment]
> the failing test is because checksum address no longer returns an result, just need to remove the unwrap
oops missed that one, thanks

> let's just keep them as clap `visible_alias`
alright

**Review by gakonst — COMMENTED** [review]
Thanks for this! Hate to be this guy but can we split the PR in separate ones that each do a specific thing? The rename/refactor of a bunch of things makes somethings a bit hard to parse I think? I love the changes but ideally we can get them in separately.

**Comment by mattsse** [comment]
I guess it would make sense to submit at least the changes/additions to `base.rs` and all the Cast improvements/updates separately.

**Comment by DaniPopes** [comment]
makes sense, closing this then
